### PR TITLE
[MXNET-422] Distributed training tutorial

### PR DESCRIPTION
**Pull request by indhub**
## Description ##
Distributed training tutorial

Note: Images might not be visible until https://github.com/dmlc/web-data/pull/67 is merged.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- [x] Code is well-documented: 
- [x] For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html

### CODE REVIEW

File: docs/tutorials/index.md [hunk 1]
```diff
@@ -38,6 +38,7 @@ Select API:&nbsp;
     * [Visual Question Answering](http://gluon.mxnet.io/chapter08_computer-vision/visual-question-answer.html) <img src="https://upload.wikimedia.org/wikipedia/commons/6/6a/External_link_font_awesome.svg" alt="External link" height="15px" style="margin: 0px 0px 3px 3px;"/>
 * Practitioner Guides
     * [Multi-GPU training](http://gluon.mxnet.io/chapter07_distributed-learning/multiple-gpus-gluon.html) <img src="https://upload.wikimedia.org/wikipedia/commons/6/6a/External_link_font_awesome.svg" alt="External link" height="15px" style="margin: 0px 0px 3px 3px;"/>
+    * [Distributed Training](https://github.com/apache/incubator-mxnet/tree/master/example/distributed_training)
```
Comment (ThomasDelteil) [inline docs/tutorials/index.md]: Not a big fan to link to the github repo. I understand the tutorial is not runnable but I would prefer if it was hosted in the tutorials/docs/gluon to avoid fragmentation. Add the right import statements, and put in 
```
runnable: 
    ```python 
    store = mxnet.kv.create('dist')
    ```
not meant to be run: 
    ```
     for batch in train_data:
         train_batch(batch, ctx, net, trainer)
    ```
```
That way the code you are using is tested against the CI for correctness.

File: example/distributed_training/README.md [hunk 2]
```diff
@@ -0,0 +1,231 @@
+# Distributed Training using Gluon
```
Comment (ThomasDelteil) [inline example/distributed_training/README.md]: Please move this to `docs/tutorials` and change this `README.md` to refer this file instead

File: example/distributed_training/README.md [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+Deep learning models are usually trained using GPUs because GPUs can do a lot more computations in parallel that CPUs. But even with the modern GPUs, it could take several days to train big models. Training can be done faster by using multiple GPUs like described in [this](https://gluon.mxnet.io/chapter07_distributed-learning/multiple-gpus-gluon.html) tutorial. However only a certain number of GPUs can be attached to one host (typically 8 or 16). To make the training even faster, we can use multiple GPUs attached to multiple hosts.
+
+In this tutorial, we will show how to train a model faster using multi-host distributed training.
+
+![Multiple GPUs connected to multiple hosts](https://raw.githubusercontent.com/dmlc/web-data/master/mxnet/example/distributed_training/distributed_training.png)
+
+We will use data parallelism to distribute the training which involves splitting the training data across GPUs attached to multiple hosts. Since the hosts are working with different subset of the training data in parallel, the training completes lot faster.
+
+In this tutorial, we will train a LeNet network using MNIST dataset using two hosts each having four GPUs.
```
Comment (ThomasDelteil) [inline example/distributed_training/README.md]: could we use CIFAR10 instead? Because multi host multi gpu for mnist is a bit overkill since it trains in 2-3 seconds on a single GPU already?

File: example/distributed_training/README.md [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+```
+python ~/mxnet/tools/launch.py -n 2 -s 2 -H hosts \
+    --sync-dst-dir /home/ubuntu/mnist_dist \
+    --launcher ssh \
+    "python /home/ubuntu/mnist_dist/mnist_dist.py"
+```
+
+- `-n 2` specifies the number of workers that must be launched
+- `-s 2` specifies the number of parameter servers that must be launched.
+- `--sync-dst-dir` specifies a destination location where the contents of the current directory with be rsync'd
```
Comment (ThomasDelteil) [inline example/distributed_training/README.md]: *with be* => will be

File: example/distributed_training/README.md [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+- `--launcher ssh` tells `launch.py` to use ssh to login to each machine in the cluster and launch processes.
```
Comment (ThomasDelteil) [inline example/distributed_training/README.md]: *login to* => login on ?
*launch processes* => launch the processes ?

File: docs/tutorials/index.md (hunk 1 above)
Comment (indhub) [reply docs/tutorials/index.md]: I don't think the CI system can currently test distributed training. I think you are trying to say part of the code can be tested. Note that it will only test very trivial parts and those parts will lose syntax highlighting which will spoils user experience.

File: example/distributed_training/README.md (hunk 2 above)
Comment (indhub) [reply example/distributed_training/README.md]: If I move this to `docs/tutorials`, I can't make it pass the tutorials tests without putting most of the code under "\`\`\`" instead of "\`\`\`python". This will remove syntax highlighting and result is bad user experience. If there is a way to whitelist this tutorial from the automated tests, please let me know.

Comment (ThomasDelteil) [reply example/distributed_training/README.md]: I only see this block not being able to run in a notebook?
```
+```python
+for batch in train_data:
+    # Train the batch using multiple GPUs
+    train_batch(batch, ctx, net, trainer)
+```
```

File: example/distributed_training/README.md [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+In distributed training (using data parallelism), training data is split into equal parts across all workers and each worker uses its subset of the training data for training. For example, if we had two machines, each running a worker, each worker managing four GPUs we'll split the data like shown below. Note that we don't split the data depending on the number of GPUs but split it depending on the number of workers.
+
+![Splitting data](https://raw.githubusercontent.com/dmlc/web-data/master/mxnet/example/distributed_training/split_data.png)
+
+Each worker can find out the total number of workers in the cluster and its own rank which is an integer between 0 and N-1 where N is the number of workers.
+
+```python
+store = kv.create('dist')
+print("Total number of workers: %d" % store.num_workers)
+print("This worker's rank: %d" % store.rank)
```
[the remark this answers is not in the document]
Comment (indhub) [reply example/distributed_training/README.md]: Few more blocks:

```store = mxnet.kv.create('dist’)```
won’t work without a bunch of environment variables being defined. launch.py defines these variables.


```trainer = gluon.Trainer(net.collect_params(),```
won’t work because net is not defined


```print("Total number of workers: %d" % store.num_workers)```
won’t work because store cannot be created. See above.

File: example/distributed_training/README.md (hunk 2 above)
Comment (indhub) [reply example/distributed_training/README.md]: Few more blocks:

```store = mxnet.kv.create('dist’)```
won’t work without a bunch of environment variables being defined. launch.py defines these variables.


```trainer = gluon.Trainer(net.collect_params(),```
won’t work because net is not defined


```print("Total number of workers: %d" % store.num_workers)```
won’t work because store cannot be created. See above.

Comment (ThomasDelteil) [reply example/distributed_training/README.md]: I see, I just don't like the idea of pointing to a github readme for a tutorial from a UI and UX perspective and also have no test coverage on it. I'll let others weight on it, but if that's the only solution... I'm just not happy with it, considering what happened with the tutorial on the straight dope etc. I wonder if we could work out a specific test case where we run a distributed training in a bunch of docker containers.

**Comment by indhub** [comment]
I'm considering testing this tutorial as part of nightly tests. While we might not have a distributed setup in the CI for testing this, it is possible to run all workers on the same machine. As far as there is enough memory in the GPU for two workers to work simultaneously, it should work and that should be good enough for testing the tutorial.

I think something like this should work:
```
python launch.py -n 2 -s 2 --launcher local python cifar10_dist.py
```

File: example/distributed_training/README.md [hunk 7]
```diff
@@ -0,0 +1,255 @@
+# Distributed Training using Gluon
+
+Deep learning models are usually trained using GPUs because GPUs can do a lot more computations in parallel that CPUs. But even with the modern GPUs, it could take several days to train big models. Training can be done faster by using multiple GPUs like described in [this](https://gluon.mxnet.io/chapter07_distributed-learning/multiple-gpus-gluon.html) tutorial. However only a certain number of GPUs can be attached to one host (typically 8 or 16). To make the training even faster, we can use multiple GPUs attached to multiple hosts.
```
Comment (eric-haibin-lin) [inline example/distributed_training/README.md]: that CPUs -> than CPUs

File: example/distributed_training/README.md (hunk 2 above)
Comment (eric-haibin-lin) [reply example/distributed_training/README.md]: If the concern is about test coverage, can we add distributed_training/cifar10_dist.py to nightly test suite, with some assertion of accuracy on validation set?

**Comment by sandeep-krishnamurthy** [comment]
@indhub - Can you update the PR with rebasing and we can get it merged? This tutorial is useful for the users.

**Comment by sandeep-krishnamurthy** [comment]
@indhub - Thanks for rebasing. Can you please comments on testing this tutorial in nightly test?

**Comment by eric-haibin-lin** [comment]
Given this is a very useful tutorial and to make distributed training tutorial testing work requires much more work, I'm merging this for now and created a issue to track this https://github.com/apache/incubator-mxnet/issues/12363
